File: src/schematic/export_svg.rs
/// Export schematic diagrams as SVG strings.
///
/// Converts a list of `SchematicElement`s into an SVG document
/// that can be saved to a file or copied to clipboard.

use crate::schematic::primitives::SchematicElement;
use std::fmt::Write;

/// Export a list of schematic elements to an SVG string.
///
/// # Arguments
/// * `elements` - List of schematic elements to render
/// * `width` - SVG canvas width in points
/// * `height` - SVG canvas height in points
///
/// # Returns
/// A complete SVG document as a String.
pub fn export_svg(elements: &[SchematicElement], width: f32, height: f32) -> String {
    let mut svg = String::new();

    writeln!(
        svg,
        r#"<?xml version="1.0" encoding="UTF-8"?>"#
    )
    .unwrap();
    writeln!(
        svg,
        r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 {w} {h}" width="{w}pt" height="{h}pt">"#,
        w = width as i32,
        h = height as i32,
    )
    .unwrap();

    // Add a white background
    writeln!(svg, r#"  <rect width="100%" height="100%" fill="white"/>"#).unwrap();

    // Default styles
    writeln!(
        svg,
        r#"  <style>
    .wire {{ stroke: black; stroke-width: 1.5; fill: none; }}
    .comp {{ stroke: black; stroke-width: 1.5; fill: none; }}
    .label {{ font-family: monospace; font-size: 11px; fill: #333; }}
    .value {{ font-family: monospace; font-size: 10px; fill: #666; }}
    .ground {{ stroke: black; stroke-width: 1.5; fill: none; }}
  </style>"#
    )
    .unwrap();

    for element in elements {
        match element {
            SchematicElement::Wire { from, to } => {
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>"#,
                    x1 = from.x,
                    y1 = from.y,
                    x2 = to.x,
                    y2 = to.y,
                )
                .unwrap();
            }
            SchematicElement::Source { pos, label, value } => {
                let cx = pos.x;
                let cy = pos.y;
                let r = 14.0;
                // Circle
                writeln!(
                    svg,
                    r#"  <circle class="comp" cx="{cx}" cy="{cy}" r="{r}"/>"#,
                    cx = cx,
                    cy = cy,
                    r = r,
                )
                .unwrap();
                // +/- signs
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x1}" y="{y1}" text-anchor="middle">+</text>"#,
                    x1 = cx - 4.0,
                    y1 = cy - 8.0 + 4.0,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x1}" y="{y1}" text-anchor="middle">-</text>"#,
                    x1 = cx - 4.0,
                    y1 = cy + 8.0 + 4.0,
                )
                .unwrap();
                // Label
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{l}: {v}</text>"#,
                    x = cx + r + 5.0,
                    y = cy + 3.0,
                    l = label,
                    v = value,
                )
                .unwrap();
            }
            SchematicElement::Resistor { pos, label, value } => {
                let cx = pos.x;
                let cy = pos.y;
                let w = 30.0;
                let h = 14.0;
                // Rectangle
                writeln!(
                    svg,
                    r#"  <rect class="comp" x="{x}" y="{y}" width="{w}" height="{h}"/>"#,
                    x = cx - w / 2.0,
                    y = cy - h / 2.0,
                    w = w,
                    h = h,
                )
                .unwrap();
                // Leads
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - w / 2.0 - 10.0,
                    x2 = cx - w / 2.0,
                    y = cy,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx + w / 2.0,
                    x2 = cx + w / 2.0 + 10.0,
                    y = cy,
                )
                .unwrap();
                // Label
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{l}: {v}</text>"#,
                    x = cx + w / 2.0 + 15.0,
                    y = cy + 3.0,
                    l = label,
                    v = value,
                )
                .unwrap();
            }
            SchematicElement::Inductor { pos, label, value } => {
                let cx = pos.x;
                let cy = pos.y;
                let segments = 4;
                let seg_w = 8.0;
                let seg_h = 10.0;
                let total_w = segments as f32 * seg_w;
                let start_x = cx - total_w / 2.0;

                // Build zigzag path
                let mut d = format!("M {:.1} {:.1}", start_x - 10.0, cy);
                for i in 0..segments {
                    let x1 = start_x + (i as f32 + 0.5) * seg_w;
                    let y1 = if i % 2 == 0 {
                        cy - seg_h
                    } else {
                        cy + seg_h
                    };
                    write!(d, " L {:.1} {:.1}", x1, y1).unwrap();
                }
                write!(d, " L {:.1} {:.1}", start_x + total_w + 10.0, cy).unwrap();

                writeln!(svg, r#"  <path class="comp" d="{d}"/>"#, d = d).unwrap();
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}" text-anchor="middle">{l}: {v}</text>"#,
                    x = cx,
                    y = cy - seg_h - 5.0,
                    l = label,
                    v = value,
                )
                .unwrap();
            }
            SchematicElement::Capacitor { pos, label, value } => {
                let cx = pos.x;
                let cy = pos.y;
                let half_plate = 10.0;

                // Two plates
                writeln!(
                    svg,
                    r#"  <line class="comp" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>"#,
                    x1 = cx - half_plate,
                    y1 = cy,
                    x2 = cx + half_plate,
                    y2 = cy,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="comp" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>"#,
                    x1 = cx - half_plate,
                    y1 = cy + 12.0,
                    x2 = cx + half_plate,
                    y2 = cy + 12.0,
                )
                .unwrap();
                // Leads
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x}" y1="{y1}" x2="{x}" y2="{y2}"/>"#,
                    x = cx,
                    y1 = cy - 8.0,
                    y2 = cy,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x}" y1="{y1}" x2="{x}" y2="{y2}"/>"#,
                    x = cx,
                    y1 = cy + 12.0,
                    y2 = cy + 20.0,
                )
                .unwrap();
                // Label
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{l}: {v}</text>"#,
                    x = cx + half_plate + 12.0,
                    y = cy + 8.0,
                    l = label,
                    v = value,
                )
                .unwrap();
            }
            SchematicElement::Diode { pos, label } => {
                let cx = pos.x;
                let cy = pos.y;

                // Triangle (anode side)
                let tri_points = format!(
                    "{:.1},{:.1} {:.1},{:.1} {:.1},{:.1}",
                    cx,
                    cy - 8.0,
                    cx,
                    cy + 8.0,
                    cx + 12.0,
                    cy
                );
                writeln!(
                    svg,
                    r#"  <polygon class="comp" points="{pts}"/>"#,
                    pts = tri_points
                )
                .unwrap();

                // Bar (cathode)
                writeln!(
                    svg,
                    r#"  <line class="comp" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>"#,
                    x1 = cx + 14.0,
                    y1 = cy - 10.0,
                    x2 = cx + 14.0,
                    y2 = cy + 10.0,
                )
                .unwrap();

                // Leads
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - 8.0,
                    x2 = cx,
                    y = cy,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx + 14.0,
                    x2 = cx + 22.0,
                    y = cy,
                )
                .unwrap();

                // Label
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{l}</text>"#,
                    x = cx + 22.0 + 12.0,
                    y = cy + 3.0,
                    l = label,
                )
                .unwrap();
            }
            SchematicElement::Switch { pos, label } => {
                let cx = pos.x;
                let cy = pos.y;

                // Lead in
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - 12.0,
                    x2 = cx,
                    y = cy,
                )
                .unwrap();

                // Switch contact (angled)
                writeln!(
                    svg,
                    r#"  <line class="comp" x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}"/>"#,
                    x1 = cx,
                    y1 = cy,
                    x2 = cx + 6.0,
                    y2 = cy - 8.0,
                )
                .unwrap();

                // Lead out
                writeln!(
                    svg,
                    r#"  <line class="wire" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx + 12.0,
                    x2 = cx + 16.0,
                    y = cy,
                )
                .unwrap();

                // Label
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}" text-anchor="middle">{l}</text>"#,
                    x = cx + 6.0,
                    y = cy - 14.0,
                    l = label,
                )
                .unwrap();
            }
            SchematicElement::Ground { pos } => {
                let cx = pos.x;
                let cy = pos.y;

                // Vertical
                writeln!(
                    svg,
                    r#"  <line class="ground" x1="{x}" y1="{y1}" x2="{x}" y2="{y2}"/>"#,
                    x = cx,
                    y1 = cy,
                    y2 = cy + 6.0,
                )
                .unwrap();

                // Three horizontal lines
                writeln!(
                    svg,
                    r#"  <line class="ground" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - 10.0,
                    x2 = cx + 10.0,
                    y = cy + 6.0,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="ground" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - 6.0,
                    x2 = cx + 6.0,
                    y = cy + 10.0,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <line class="ground" x1="{x1}" y1="{y}" x2="{x2}" y2="{y}"/>"#,
                    x1 = cx - 3.0,
                    x2 = cx + 3.0,
                    y = cy + 14.0,
                )
                .unwrap();
            }
            SchematicElement::Node { pos, label } => {
                let cx = pos.x;
                let cy = pos.y;
                writeln!(
                    svg,
                    r#"  <circle class="comp" cx="{cx}" cy="{cy}" r="3" fill="black"/>"#,
                    cx = cx,
                    cy = cy,
                )
                .unwrap();
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{l}</text>"#,
                    x = cx + 6.0,
                    y = cy + 3.0,
                    l = label,
                )
                .unwrap();
            }
            SchematicElement::Label { pos, text } => {
                writeln!(
                    svg,
                    r#"  <text class="label" x="{x}" y="{y}">{t}</text>"#,
                    x = pos.x,
                    y = pos.y + 3.0,
                    t = text,
                )
                .unwrap();
            }
        }
    }

    writeln!(svg, "</svg>").unwrap();
    svg
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::schematic::primitives::Pos;
    use crate::schematic::primitives::SchematicElement::*;

    #[test]
    fn test_export_svg_basic() {
        let elements = vec![
            Wire {
                from: Pos::new(10.0, 10.0),
                to: Pos::new(100.0, 10.0),
            },
            Ground {
                pos: Pos::new(50.0, 30.0),
            },
            Label {
                pos: Pos::new(10.0, 40.0),
                text: "Test".to_owned(),
            },
        ];

        let svg = export_svg(&elements, 200.0, 100.0);

        assert!(svg.starts_with("<?xml"));
        assert!(svg.contains("<svg"));
        assert!(svg.contains("</svg>"));
        assert!(svg.contains("Test"));
        assert!(svg.contains("white"));
    }

    #[test]
    fn test_export_svg_source() {
        let elements = vec![Source {
            pos: Pos::new(50.0, 50.0),
            label: "Vin".to_owned(),
            value: "12 V".to_owned(),
        }];

        let svg = export_svg(&elements, 200.0, 100.0);
        assert!(svg.contains("Vin"));
        assert!(svg.contains("12 V"));
    }
}

File: src/schematic/ExportSvg.cpp
#include "ExportSvg.h"
#include <sstream>
#include <string>
#include <vector>

namespace export_svg {

static std::string fmt(float v) {
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%.1f", v);
    return buf;
}

static void write_wire(std::ostringstream& svg, const SchematicWire& wire) {
    svg << "  <line class=\"wire\" x1=\"" << fmt(wire.from.x)
        << "\" y1=\"" << fmt(wire.from.y)
        << "\" x2=\"" << fmt(wire.to.x)
        << "\" y2=\"" << fmt(wire.to.y) << "\"/>\n";
}

static void write_source(std::ostringstream& svg, const SchematicSource& src) {
    float cx = src.pos.x;
    float cy = src.pos.y;
    float r = 14.0f;

    svg << "  <circle class=\"comp\" cx=\"" << fmt(cx) << "\" cy=\"" << fmt(cy) << "\" r=\"" << fmt(r) << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx - 4.0f) << "\" y=\"" << fmt(cy - 8.0f + 4.0f) << "\" text-anchor=\"middle\">+</text>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx - 4.0f) << "\" y=\"" << fmt(cy + 8.0f + 4.0f) << "\" text-anchor=\"middle\">-</text>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + r + 5.0f) << "\" y=\"" << fmt(cy + 3.0f)
        << "\">" << src.label << ": " << src.value << "</text>\n";
}

static void write_resistor(std::ostringstream& svg, const SchematicResistor& res) {
    float cx = res.pos.x;
    float cy = res.pos.y;
    float w = 30.0f;
    float h = 14.0f;

    svg << "  <rect class=\"comp\" x=\"" << fmt(cx - w / 2.0f)
        << "\" y=\"" << fmt(cy - h / 2.0f)
        << "\" width=\"" << fmt(w) << "\" height=\"" << fmt(h) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx - w / 2.0f - 10.0f)
        << "\" y1=\"" << fmt(cy) << "\" x2=\"" << fmt(cx - w / 2.0f)
        << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx + w / 2.0f)
        << "\" y1=\"" << fmt(cy) << "\" x2=\"" << fmt(cx + w / 2.0f + 10.0f)
        << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + w / 2.0f + 15.0f)
        << "\" y=\"" << fmt(cy + 3.0f)
        << "\">" << res.label << ": " << res.value << "</text>\n";
}

static void write_inductor(std::ostringstream& svg, const SchematicInductor& ind) {
    float cx = ind.pos.x;
    float cy = ind.pos.y;
    int segments = 4;
    float seg_w = 8.0f;
    float seg_h = 10.0f;
    float total_w = segments * seg_w;
    float start_x = cx - total_w / 2.0f;

    std::string d = "M " + fmt(start_x - 10.0f) + " " + fmt(cy);
    for (int i = 0; i < segments; ++i) {
        float x1 = start_x + (i + 0.5f) * seg_w;
        float y1 = (i % 2 == 0) ? (cy - seg_h) : (cy + seg_h);
        d += " L " + fmt(x1) + " " + fmt(y1);
    }
    d += " L " + fmt(start_x + total_w + 10.0f) + " " + fmt(cy);

    svg << "  <path class=\"comp\" d=\"" << d << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx)
        << "\" y=\"" << fmt(cy - seg_h - 5.0f)
        << "\" text-anchor=\"middle\">" << ind.label << ": " << ind.value << "</text>\n";
}

static void write_capacitor(std::ostringstream& svg, const SchematicCapacitor& cap) {
    float cx = cap.pos.x;
    float cy = cap.pos.y;
    float half_plate = 10.0f;

    svg << "  <line class=\"comp\" x1=\"" << fmt(cx - half_plate)
        << "\" y1=\"" << fmt(cy) << "\" x2=\"" << fmt(cx + half_plate)
        << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"comp\" x1=\"" << fmt(cx - half_plate)
        << "\" y1=\"" << fmt(cy + 12.0f) << "\" x2=\"" << fmt(cx + half_plate)
        << "\" y2=\"" << fmt(cy + 12.0f) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx) << "\" y1=\"" << fmt(cy - 8.0f)
        << "\" x2=\"" << fmt(cx) << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx) << "\" y1=\"" << fmt(cy + 12.0f)
        << "\" x2=\"" << fmt(cx) << "\" y2=\"" << fmt(cy + 20.0f) << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + half_plate + 5.0f)
        << "\" y=\"" << fmt(cy + 8.0f)
        << "\">" << cap.label << ": " << cap.value << "</text>\n";
}

static void write_diode(std::ostringstream& svg, const SchematicDiode& diode) {
    float cx = diode.pos.x;
    float cy = diode.pos.y;

    svg << "  <polygon class=\"comp\" points=\""
        << fmt(cx) << "," << fmt(cy - 8.0f) << " "
        << fmt(cx) << "," << fmt(cy + 8.0f) << " "
        << fmt(cx + 12.0f) << "," << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"comp\" x1=\"" << fmt(cx + 14.0f)
        << "\" y1=\"" << fmt(cy - 10.0f)
        << "\" x2=\"" << fmt(cx + 14.0f)
        << "\" y2=\"" << fmt(cy + 10.0f) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx - 8.0f)
        << "\" y1=\"" << fmt(cy) << "\" x2=\"" << fmt(cx)
        << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx + 14.0f)
        << "\" y1=\"" << fmt(cy) << "\" x2=\"" << fmt(cx + 22.0f)
        << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + 22.0f + 5.0f)
        << "\" y=\"" << fmt(cy + 3.0f)
        << "\">" << diode.label << "</text>\n";
}

static void write_switch(std::ostringstream& svg, const SchematicSwitch& sw) {
    float cx = sw.pos.x;
    float cy = sw.pos.y;

    svg << "  <line class=\"wire\" x1=\"" << fmt(cx - 12.0f) << "\" y1=\"" << fmt(cy)
        << "\" x2=\"" << fmt(cx) << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <line class=\"comp\" x1=\"" << fmt(cx) << "\" y1=\"" << fmt(cy)
        << "\" x2=\"" << fmt(cx + 6.0f) << "\" y2=\"" << fmt(cy - 8.0f) << "\"/>\n";
    svg << "  <line class=\"wire\" x1=\"" << fmt(cx + 12.0f) << "\" y1=\"" << fmt(cy)
        << "\" x2=\"" << fmt(cx + 16.0f) << "\" y2=\"" << fmt(cy) << "\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + 6.0f) << "\" y=\"" << fmt(cy - 14.0f)
        << "\" text-anchor=\"middle\">" << sw.label << "</text>\n";
}

static void write_ground(std::ostringstream& svg, const SchematicGround& gnd) {
    float cx = gnd.pos.x;
    float cy = gnd.pos.y;
    svg << "  <line class=\"ground\" x1=\"" << fmt(cx) << "\" y1=\"" << fmt(cy)
        << "\" x2=\"" << fmt(cx) << "\" y2=\"" << fmt(cy + 6.0f) << "\"/>\n";
    svg << "  <line class=\"ground\" x1=\"" << fmt(cx - 10.0f) << "\" y1=\"" << fmt(cy + 6.0f)
        << "\" x2=\"" << fmt(cx + 10.0f) << "\" y2=\"" << fmt(cy + 6.0f) << "\"/>\n";
    svg << "  <line class=\"ground\" x1=\"" << fmt(cx - 6.0f) << "\" y1=\"" << fmt(cy + 10.0f)
        << "\" x2=\"" << fmt(cx + 6.0f) << "\" y2=\"" << fmt(cy + 10.0f) << "\"/>\n";
    svg << "  <line class=\"ground\" x1=\"" << fmt(cx - 3.0f) << "\" y1=\"" << fmt(cy + 14.0f)
        << "\" x2=\"" << fmt(cx + 3.0f) << "\" y2=\"" << fmt(cy + 14.0f) << "\"/>\n";
}

static void write_node(std::ostringstream& svg, const SchematicNode& node) {
    float cx = node.pos.x;
    float cy = node.pos.y;
    svg << "  <circle class=\"comp\" cx=\"" << fmt(cx) << "\" cy=\"" << fmt(cy)
        << "\" r=\"3\" fill=\"black\"/>\n";
    svg << "  <text class=\"label\" x=\"" << fmt(cx + 6.0f)
        << "\" y=\"" << fmt(cy + 3.0f)
        << "\">" << node.label << "</text>\n";
}

static void write_label(std::ostringstream& svg, const SchematicLabel& lbl) {
    svg << "  <text class=\"label\" x=\"" << fmt(lbl.pos.x)
        << "\" y=\"" << fmt(lbl.pos.y + 3.0f)
        << "\">" << lbl.text << "</text>\n";
}

std::string export_svg(const std::vector<SchematicElement>& elements, float width, float height) {
    std::ostringstream svg;

    svg << R"(<?xml version="1.0" encoding="UTF-8"?>)" << "\n";
    svg << "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 "
        << static_cast<int>(width) << " " << static_cast<int>(height)
        << "\" width=\"" << static_cast<int>(width) << "pt\" height=\""
        << static_cast<int>(height) << "pt\">\n";

    svg << R"(  <rect width="100%" height="100%" fill="white"/>)" << "\n";

    svg << R"(  <style>
    .wire { stroke: black; stroke-width: 1.5; fill: none; }
    .comp { stroke: black; stroke-width: 1.5; fill: none; }
    .label { font-family: monospace; font-size: 11px; fill: #333; }
    .value { font-family: monospace; font-size: 10px; fill: #666; }
    .ground { stroke: black; stroke-width: 1.5; fill: none; }
  </style>)" << "\n";

    for (const auto& elem : elements) {
        switch (elem.type) {
            case SchematicElementType::Wire:      write_wire(svg, elem.wire); break;
            case SchematicElementType::Source:    write_source(svg, elem.source); break;
            case SchematicElementType::Resistor:  write_resistor(svg, elem.resistor); break;
            case SchematicElementType::Inductor:  write_inductor(svg, elem.inductor); break;
            case SchematicElementType::Capacitor: write_capacitor(svg, elem.capacitor); break;
            case SchematicElementType::Diode:     write_diode(svg, elem.diode); break;
            case SchematicElementType::Switch:    write_switch(svg, elem.switch_); break;
            case SchematicElementType::Ground:    write_ground(svg, elem.ground); break;
            case SchematicElementType::Node:      write_node(svg, elem.node); break;
            case SchematicElementType::Label:     write_label(svg, elem.label); break;
        }
    }

    svg << "</svg>\n";
    return svg.str();
}

} // namespace export_svg
